src/wallet/sqlite.cpp

```diff
@@ -80,6 +81,72 @@ void SQLiteDatabase::Cleanup()
 
 void SQLiteDatabase::Open(const char* mode)
 {
+    int flags = SQLITE_OPEN_FULLMUTEX | SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE;
+    if (m_mock) {
+        flags |= SQLITE_OPEN_MEMORY; // In memory database for mock db
+    }
+
+    if (m_db == nullptr) {
+        TryCreateDirectories(m_dir_path);
+        int ret = sqlite3_open_v2(m_file_path.c_str(), &m_db, flags, nullptr);
+        if (ret != SQLITE_OK) {
+            throw std::runtime_error(strprintf("SQLiteDatabase: Failed to open database: %s\n", sqlite3_errstr(ret)));
+        }
+    }
+
+    if (sqlite3_db_readonly(m_db, "main") != 0) {
+        throw std::runtime_error("SQLiteDatabase: Database opened in readonly mode but read-write permissions are needed");
+    }
+
+    // Acquire an exclusive lock on the database
+    // First change the locking mode to exclusive
+    int ret = sqlite3_exec(m_db, "PRAGMA locking_mode = exclusive", nullptr, nullptr, nullptr);
+    if (ret != SQLITE_OK) {
+        throw std::runtime_error(strprintf("SQLiteDatabase: Unable to change database locking mode to exclusive: %s\n", sqlite3_errstr(ret)));
+    }
+    // Now begin a transaction to acquire the exclusive lock. This lock won't be released until we close because of the exclusive locking mode.
+    ret = sqlite3_exec(m_db, "BEGIN EXCLUSIVE TRANSACTION", nullptr, nullptr, nullptr);
+    if (ret != SQLITE_OK) {
+        throw std::runtime_error("SQLiteDatabase: Unable to obtain an exclusive lock on the database, is it being used by another bitcoind?\n");
+    }
+    ret = sqlite3_exec(m_db, "COMMIT", nullptr, nullptr, nullptr);
+    if (ret != SQLITE_OK) {
+        throw std::runtime_error(strprintf("SQLiteDatabase: Unable to end exclusive lock transaction: %s\n", sqlite3_errstr(ret)));
+    }
+
+    // Make the table for our key-value pairs
+    // First check that the main table exists
+    bool table_exists;
```

Comment:
6720df53bd5d4e551c101743c4050bd0c7682238
nit: Mind moving the `table_exists` declaration 9 lines down to its first usage?

Reply:
Done